### PR TITLE
Instruction Class Clean-Up

### LINKED ISSUE #355: Clean Up Instruction Class
In its current state, the Instruction class contains un-used functions, lacks common member variables in the abstract root class, and has a bloated implementation of storing metadata such as `isFloat` etc. This should be streamlined to improve readability and usability of the class

### PR DESCRIPTION
**Pull request by FinnWilkinson**
Generally tidies the instruction classes and moves some common functions & member variables into the base class.

Also includes new classes in `Container.hh` which mitigates slowdown caused by changing `sourceValues_`, `results_`, `sourceRegisters_`, `destinationRegisters_` to std::vector when SME was first supported. Said class has uses std::variant which contains std::array (used by default) and std::vector (used on call of `makeSME()`). This means most instructions can use std::array and get speed benefits, and SME instructions get the flexibility of std::vecotr (needed due to variable and large amount of operands).

Closes #355 

AArch64 speeds with new Constants.hh (run on M1 MacBook Pro). TX2.yaml used for tx2 binaries, a64fx.yaml for armv8.4-a(+sve) binaries, a64fx_SME.yaml for all SME binaries. SME binaries run matmul kernel for 100 iterations on square inputs.

| Binary | MIPS `dev` | time/ms `dev` | MIPS `insn-class-clean` | time/ms `insn-class-clean`|
|--------|-------------|---------------|-------------------------|--------------------------|
|bude-armclang20.0-armv8.4-a |1.2|27033|1.4|22432|
|bude-armclang20.0-armv8.4-a+sve |0.87|12522|1.1|10078|
|bude-armclang20.0-tx2 |1.2|27276|1.3|25262|
|bude-gcc10.3.0-armv8.4-a |1.2|27048|1.2|26280|
|bude-gcc10.3.0-armv8.4-a+sve |0.86|11912|0.92|11195|
|bude-gcc10.3.0-tx2 |1.2|26344|1.3|24695|
|bude-gcc8.3.0-armv8.4-a |1.1|28607|1.2|26451|
|bude-gcc8.3.0-armv8.4-a+sve |0.86|12303|0.97|10880|
|bude-gcc8.3.0-tx2 |1.2|27657|1.3|25550|
|bude-gcc9.3.0-armv8.4-a |1.1|28521|1.2|26678|
|bude-gcc9.3.0-armv8.4-a+sve |0.8|12520|0.94|10590|
|bude-gcc9.3.0-tx2 |1.2|27505|1.3|26040|
|stream-armclang20.0-armv8.4-a |0.78|16179|1|12644|
|stream-armclang20.0-armv8.4-a+sve |0.82|9180|0.99|7610|
|stream-armclang20.0-tx2 |0.64|15039|0.76|12591|
|stream-gcc10.3.0-armv8.4-a |0.97|15709|1.2|12856|
|stream-gcc10.3.0-armv8.4-a+sve |0.68|11046|0.85|8809|
|stream-gcc10.3.0-tx2 |1|14667|1.2|13210|
|stream-gcc8.3.0-armv8.4-a |1|14808|1.1|13695|
|stream-gcc8.3.0-armv8.4-a+sve |0.68|10956|0.76|9871|
|stream-gcc8.3.0-tx2 |0.99|15327|1.1|13459|
|stream-gcc9.3.0-armv8.4-a |1|14642|1.1|13574|
|stream-gcc9.3.0-armv8.4-a+sve |0.72|10311|0.75|9877|
|stream-gcc9.3.0-tx2 |1|14942|1.1|13881|
|sme_matmul_fp32 m=n=k=64 |0.72|2138|0.75|2057|
|sme_matmul_fp32 m=n=k=128 |0.8|10099|0.85|9561|
|sme_matmul_fp32 m=n=k=256 |0.82|59515|0.86|56645|

### CODE REVIEW

File: src/include/simeng/arch/riscv/Instruction.hh
Comment (jj16791) [inline src/include/simeng/arch/riscv/Instruction.hh:213]: Why have `isAtomic` and `isFloat` moved to private?

File: src/lib/arch/riscv/Instruction.cc
Comment (jj16791) [inline src/lib/arch/riscv/Instruction.cc:179]: If we move back into public, we should bump up `isAtomic` and `isFloat` in file to be with the other instruction identifier getters

**Review by jj16791 — CHANGES_REQUESTED** [review]

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Instruction.hh:225]: These comments were copied from what was there before. Not sure what they could be updated to to be more descriptive... any suggestions?

File: src/include/simeng/arch/riscv/Instruction.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/riscv/Instruction.hh:213]: They are only used internally in the instruction class, but they can probably be removed completely

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (jj16791) [reply src/include/simeng/arch/aarch64/Instruction.hh:226]: Reading this, "Operates on" doesn't map very well to our group assignment. Some instructions "operate" on two types of data (e.g. vectors and scalars) but only have one assignment re data type identifier. The comments should be changed to reflect this

File: src/include/simeng/arch/aarch64/Container.hh
Comment (jj16791) [inline src/include/simeng/arch/aarch64/Container.hh:47]: Is the intended usage of `resize` to trim vectors of unused indexes initialised due to the ADDITIONAL_SME_* and MAX_* variables?

**Review by ABenC377 — CHANGES_REQUESTED** [review]
Looks good other than a comment typo

File: src/include/simeng/arch/aarch64/Container.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Container.hh:53]: nodiscard forces the return value to be used -- if it is not used then it is a programming error

File: src/include/simeng/arch/aarch64/Container.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Container.hh:78]: I agree -- I'm unsure whether something can be used in the template to pass in what the array size should be, but I will try this out. Typically the array size must be const and known at compile time

File: src/include/simeng/arch/aarch64/Container.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Container.hh:54]: Given we are only capturing `idx` by value (as `var_` is passed by reference through `auto && arg`) passing by reference gives no advantage as the pointer size would be equal to the size of `size_t`

File: src/lib/arch/aarch64/Instruction.cc
Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction.cc:75]: Good spot - will remove

File: src/include/simeng/arch/aarch64/Container.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Container.hh:47]: Yes, it is probably not strictly needed but will make the vector only as big as it needs to be

**Review by JosephMoore25 — COMMENTED** [review]
Looks good, and yields impressive speedups. Happy to approve once other's comments have been handled, but nothing extra that I could find.

File: src/include/simeng/arch/aarch64/Container.hh
Comment (jj16791) [reply src/include/simeng/arch/aarch64/Container.hh:54]: Okay good, just double checking.

A question more than a requested change. Why do we pass var_ as an `&&` ?

File: src/include/simeng/arch/aarch64/Container.hh
Comment (jj16791) [reply src/include/simeng/arch/aarch64/Container.hh:47]: That's fine, just checking the intended use cases.

To be more direct, I think the comment would benefit from a rewording to something like "Resize the vector to be the same size as `numRegs`. Used to ensure any unused vector indexes introduced in makeSME() are removed."

**Review by JosephMoore25 — DISMISSED** [review]
Nice work. I like the change to using just `operandContainer`.

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (jj16791) [inline src/include/simeng/arch/aarch64/Instruction.hh:20]: Brief comment above these to explain their purpose

File: src/include/simeng/arch/aarch64/Container.hh
Comment (jj16791) [inline src/include/simeng/arch/aarch64/Container.hh:1]: I think this file would be better named `operandContainer` now. I don't see the need for a more generalised name given the single class. If this does get carried out, then maybe wait until all other reviewers approve as it'll show up as a whole new file to review on GitHub.

**Review by jj16791 — CHANGES_REQUESTED** [review]

File: src/include/simeng/arch/aarch64/Container.hh
Comment (FinnWilkinson) [reply src/include/simeng/arch/aarch64/Container.hh:54]: Was copying the examples from the c++ specification https://en.cppreference.com/w/cpp/utility/variant/visit

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (jj16791) [inline src/include/simeng/arch/aarch64/Instruction.hh:20]: *Operand type aliases ?

**Review by jj16791 — CHANGES_REQUESTED** [review]

**Review by jj16791 — APPROVED** [review]

**Review by ABenC377 — DISMISSED** [review]

File: src/lib/arch/aarch64/Instruction.cc
Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction.cc:49]: Not really no - we have opened an issue to better address zero registers, and the `destinationRegisterCount` must be used given the usage of the new `operandContainer` class

File: src/lib/arch/aarch64/Instruction_execute.cc
Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: It was thought that combining the booleans into a single value would tidy up the instruction class and better enforce a "getter-setter" methodology surrounding said instruction identifiers.

I wouldn't agree on reduced readability personally - open for input from the others on this though @jj16791 @JosephMoore25 @ABenC377

File: src/lib/arch/riscv/Instruction_address.cc
Comment (FinnWilkinson) [reply src/lib/arch/riscv/Instruction_address.cc:16]: `isLoad` and `isStoreAddress` are public functions required by the base class and their usage is left over from previous implementation (isAtomic was replaced using find-replace). Given the new `isInstruction` function, and given it was only used inside the instruction class,`isAtomic` was removed as it was seen as redundant

File: src/lib/arch/aarch64/Instruction_execute.cc
Comment (jj16791) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: My preference is for the changes made here as it provides a solution that looks cleaner than a set of boolean getters/setters. I also second Finn's comment regarding "reduced readability". Prior offline analysis has also shown no performance degradation from this change.

Personally, I also find this solution nicer when adding new instruction identifiers than the previous one.

Comment (dANW34V3R) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: I get the point about the use of getters and setters as we previously mixed them with directly accessing the member variable which is not ideal. It seems to me that the a solution of enforcing the previous getters and setters would be better. i.e. we never directly use the 'float_' variable.  Although that does mean we have lots of extra functions in the instruction class class. OR we only use the member variables and don't use getters and setters, they are all private anyway only exposed through the parent class.

Is there a limit on the size of 'isSVEDataMask' etc? And therefor the number of different identifiers we can have? I guess 'int' and therefore 64?

Comment (dANW34V3R) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: On readability it is this

`isScalarData_ || isVectorData_ || isSVEData_
`
vs this

```
isInstruction(InsnIdentifier::isScalarDataMask) ||
                            isInstruction(InsnIdentifier::isVectorDataMask) ||
                            isInstruction(InsnIdentifier::isSVEDataMask))
```

Comment (ABenC377) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: I think both approaches are entirely readable and there doesn't seem to be any performance delta, so would plumb for whichever approach is more consistent with our overall methodology.  I'm convinced by the "getter-setter" methodology argument and so vote for changes to be kept.

Comment (ABenC377) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: The second one is longer, but I don't think it is any more difficult to understand.  Not sure on our position on verbosity per se is. 

> On readability it is this
> 
> `isScalarData_ || isVectorData_ || isSVEData_ ` vs this
> 
> ```
> isInstruction(InsnIdentifier::isScalarDataMask) ||
>                             isInstruction(InsnIdentifier::isVectorDataMask) ||
>                             isInstruction(InsnIdentifier::isSVEDataMask))
> ```

Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: The upper limit for the number of masks would be 64, however with AArch64 currently at 15 masks I do not think we are in danger of reaching this limit for current or future architectures.

If we enforced the getter-setter philosophy with the previous implementation, it would mean for each new identifier we would have to define two new functions. For the new implementation it requires just adding the mask.
(However, given, as Dan said, we previously just accessed the member variables directly in the instruction class this was less of an issue previously.)

File: src/lib/arch/riscv/Instruction_address.cc
Comment (FinnWilkinson) [reply src/lib/arch/riscv/Instruction_address.cc:16]: If preferable, `isLoad` and `isStoreAddress` can be changes to `isInstruction` to maintain consistent semantics

File: src/lib/arch/aarch64/Instruction_execute.cc
Comment (JosephMoore25) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: I've given this a bit of a think. I like the bitfield a lot more than separate bools so lean towards the proposed implementation, but I do agree with Dan where it reduces readability.
I tried to have a think of other solutions, but I think the best option would be stick with this solution and change a couple of things.
- We can OR each mask before passing into isInstruction, this leads to a single call (less syntax) and doesn't require any functionality changes for it to still work
- Remove `DataMask` from the variable names. Given each bit acts as a bool, I think `isVector` for example is descriptive enough, especially when bundled with the Enum name
- Reduce `InsnIdentifier` to `InsnType`. Therefore also `setInstructionIdentifier` to `setInstructionType`
These changes dont change functionality and I believe maintain the same level of clarity, but make some snippets of the code a bit nicer.

It'd lead to the above snippet being:
```
if (!(isInstruction(InsnType::isScalar || 
		    InsnType::isVector ||
		    InsnType::isSVE)) {
      setInstructionType(InsnType::isScalar);
}
```
I think this is the best middleground achievable in this, as I agree with points on both sides.

Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: @JosephMoore25 I think these are all good suggestions, especially OR-ing the masks before passing into `isInstruction`, and provide a good balance between new and old. Although, the use of `data` in some of the current masks gives clarity as to what they represent so will just remove the `mask` suffix

File: src/include/simeng/arch/aarch64/helpers/auxiliaryFunctions.hh
Comment (dANW34V3R) [inline src/include/simeng/arch/aarch64/helpers/auxiliaryFunctions.hh:245]: I think this name change needs to be applied to the docs in aarch64.rst

File: src/lib/arch/aarch64/Instruction_execute.cc
Comment (JosephMoore25) [reply src/lib/arch/aarch64/Instruction_execute.cc:85]: @tom91136 is correct and that's ruined my morning :( . If we still want my solution, `InsnType` would need to be an `enum` rather than `enum class`, or we could overload the | operator which would be a ~3 line function. Not ideal, but imo still improves clarity (although an argument can be made of how far do you go to improve clarity...)

Also, small mistake in my code snippet where we should be `|`ing the masks, not `||`ing them.

File: src/lib/arch/aarch64/Instruction.cc
Comment (JosephMoore25) [inline src/lib/arch/aarch64/Instruction.cc:175]: Shouldn't this be:
`if (isInsnOneOf(InsnType::isLoad, InsnType::isStoreAddress)) {`?

**Review by JosephMoore25 — CHANGES_REQUESTED** [review]
Looks good. I appreciate the new changes to what was `isInstruction` to make the general syntax of the Instruction files nicer. I think you missed one case where we can apply the new function better, but other than that all good.

File: src/lib/arch/aarch64/Instruction.cc
Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction.cc:175]: Good spot

**Review by JosephMoore25 — DISMISSED** [review]

**Review by ABenC377 — DISMISSED** [review]

File: src/lib/arch/aarch64/Instruction_execute.cc
Comment (JosephMoore25) [reply src/lib/arch/aarch64/Instruction_execute.cc]: A different implementation to the one suggested above was put in place. I'm now satisfied this issue is resolved, but will leave to @dANW34V3R to resolve this comment chain as they raised this point in the first place.

File: src/lib/arch/riscv/Instruction_address.cc
Comment (FinnWilkinson) [reply src/lib/arch/riscv/Instruction_address.cc]: This has been changed so that `isLoad()` and `isStoreAddress()` are not used

**Review by ABenC377 — DISMISSED** [review]
👏🏻

**Review by jj16791 — CHANGES_REQUESTED** [review]

File: test/unit/aarch64/OperandContainerTest.cc
Comment (FinnWilkinson) [reply test/unit/aarch64/OperandContainerTest.cc:19]: An assert and debug-dependant testing can be added to help future debugging if this is called on a non-vector

**Review by ABenC377 — APPROVED** [review]

**Review by JosephMoore25 — DISMISSED** [review]
Looks good. A shame we couldn't get isInstruction cleaner, but I'll have a think for next release.

**Review by dANW34V3R — DISMISSED** [review]

**Review by jj16791 — DISMISSED** [review]

**Review by JosephMoore25 — DISMISSED** [review]

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (dANW34V3R) [inline src/include/simeng/arch/aarch64/Instruction.hh:417]: This isn't really a vector of source registers anymore. Comment needs updating

File: src/include/simeng/arch/aarch64/Instruction.hh
Comment (dANW34V3R) [inline src/include/simeng/arch/aarch64/Instruction.hh:429]: Comment updates here also

**Review by JosephMoore25 — APPROVED** [review]

**Review by jj16791 — APPROVED** [review]

**Review by dANW34V3R — APPROVED** [review]

**Pull request merged**
